File: src/#173/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val
  this.left = left === undefined ? null : left
  this.right = right === undefined ? null : right
}

var BSTIterator = function (root) {
  let arr = []

  function inorderTraversal(root) {
    if (!root) {
      return
    }
    inorderTraversal(root.left)
    arr.push(root.val)
    inorderTraversal(root.right)
  }
  inorderTraversal(root)
  this.inorderArr = arr
  this.index = 0
}

/**
 * @return the next smallest number
 * @return {number}
 */
BSTIterator.prototype.next = function () {
  if (this.index < this.inorderArr.length) {
    let res = this.inorderArr[this.index]
    this.index++
    return res
  }
  return 0
}

/**
 * @return whether we have a next smallest number
 * @return {boolean}
 */
BSTIterator.prototype.hasNext = function () {
  if (this.index < this.inorderArr.length) {
    return true
  }
  return false
}

File: src/#173/solution.go
package solution

type TreeNode struct {
	Val         int
	Left, Right *TreeNode
}

type BSTIterator struct {
	InorderArr []int
	Index      int
}

func Constructor(root *TreeNode) BSTIterator {
	arr := []int{}
	inorderTraversal(root, &arr)
	return BSTIterator{
		InorderArr: arr,
		Index:      0,
	}
}

/** @return the next smallest number */
func (this *BSTIterator) Next() int {
	if this.Index < len(this.InorderArr) {
		res := this.InorderArr[this.Index]
		this.Index++
		return res
	}
	return 0
}

/** @return whether we have a next smallest number */
func (this *BSTIterator) HasNext() bool {
	if this.Index < len(this.InorderArr) {
		return true
	}
	return false
}

func inorderTraversal(root *TreeNode, arr *[]int) {
	if root == nil {
		return
	}
	inorderTraversal(root.Left, arr)
	*arr = append(*arr, root.Val)
	inorderTraversal(root.Right, arr)
}
